crypto/evp/m_sigver.c

```diff
@@ -43,59 +44,74 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
         ctx->provctx = NULL;
     }
 
-    if (ctx->pctx == NULL) {
+    if (ctx->pctx == NULL)
         ctx->pctx = EVP_PKEY_CTX_new(pkey, e);
-        if (ctx->pctx == NULL)
-            return 0;
-    } else if (pkey != NULL) {
-        if (!EVP_PKEY_up_ref(pkey))
-            return 0;
-        EVP_PKEY_free(ctx->pctx->pkey);
-        ctx->pctx->pkey = pkey;
-    }
+    if (ctx->pctx == NULL)
+        return 0;
+
     locpctx = ctx->pctx;
     evp_pkey_ctx_free_old_ops(locpctx);
-    if (locpctx->pkey == NULL)
-        goto legacy;
 
-    if (e != NULL || locpctx->engine != NULL)
+    if (locpctx->algorithm == NULL)
         goto legacy;
 
-    if (signature != NULL) {
-        if (!EVP_SIGNATURE_up_ref(signature))
-            goto err;
-    } else {
-        /*
-         * TODO(3.0): Check for legacy handling. Remove this once all all
-         * algorithms are moved to providers.
-         */
-        switch (locpctx->pkey->type) {
-        case NID_dsa:
-            break;
-        default:
-            goto legacy;
+    if (mdname == NULL) {
+        if (type != NULL) {
+            mdname = EVP_MD_name(type);
+        } else if (pkey != NULL) {
+            /*
+             * TODO(v3.0) work out a better way for EVP_PKEYs with no legacy
+             * component.
+             */
+            if (pkey->pkey.ptr != NULL) {
+                int def_nid;
+                if (EVP_PKEY_get_default_digest_nid(pkey, &def_nid) > 0)
+                    mdname = OBJ_nid2sn(def_nid);
+            }
         }
-        signature
-            = EVP_SIGNATURE_fetch(NULL, OBJ_nid2sn(locpctx->pkey->type), NULL);
+    }
```

Comment:
Do we need to handle the cases in which after this block mdname is still null?

E.g. mdname, type and pkey are NULL or no default digest nid, etc.

Reply:
Ah, good question...  If I understand correctly, we've left that for the provided init functions to decide.  @mattcaswell, please confirm.

Originally, I think a NULL `type` was allowed to accommodate contraptions like MACs implemented as PKEY signature methods, which can obviously take any size input.  However, that doesn't exclude the possibility that some public key algos might be able to sign any size data without an intermediary hash function, now or in the future, so it may still be wise to let the provided init functions decide.  After all, the EVP API is supposed to be a fairly thin layer, let's try and keep it that way.

Reply:
However, I just noticed that `mdname` is processed a second time in lines 120+.  I need to have a closer look at that.

Reply:
This function is called as a result of the application calling `EVP_DigestSignInit` or `EVP_DigestSignInit_ex`. In the former case we pass through the EVP_MD from the application in the `type` parameter. Otherwise we pass through the `mdname`. So, normally, one or the other is set. The only case where they are both NULL would be if the application explicitly passed a NULL - in which case we should just pass that through to the provider to interpret.

Reply:
> However, I just noticed that mdname is processed a second time in lines 120+. I need to have a closer look at that.

That second processing doesn't quite look right...

Reply:
> That second processing doesn't quite look right...

It isn't right at all...  it's a remnant that I missed.  I'll fix.